runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -136,32 +135,33 @@ public synchronized void acquireBuffer() {
 
     @Override
     public int getSizeEstimate() {
-        byte[] tempData = data;
-        if (tempData != null) {
-            return tempData.length;
-        }
-
         if (lastKnownSize != NOT_KNOWN) {
             return lastKnownSize;
         }
-
         return 1;
     }
 
     @Getter
-    final EnumMap<LogUnitMetadataType, Object> metadataMap;
+    EnumMap<LogUnitMetadataType, Object> metadataMap;
+    private ByteBuf serializedMetadata = null;
+
+    public void serializeMetadata() {
+        serializedMetadata = Unpooled.buffer();
+        ICorfuPayload.serialize(serializedMetadata, metadataMap);
+        metadataMap = null;
+    }
 
     /**
      * Return the payload.
      */
     public LogData(ByteBuf buf) {
+        lastKnownSize = buf.readableBytes();
```

Comment:
should use buf.readerIndex()

Reply:
This is a fresh buffer, so readableBytes() should be correct?  Let me run a couple of tests...

Reply:
It is not guarantee a fresh buffer. As runtime ask for log data to the server, the LogData will be sent in a batch in one message. The message payload buffer contains a batch of LogData.